// @crystalui/angular-lightbox/esm2015/lib/lightbox.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver, ApplicationRef } from '@angular/core';
import { LightboxComponent } from './lightbox.component';
import { DefaultProperties } from './default-properties';
/**
 * @record
 */
export function AdComponent() { }
if (false) {
    /** @type {?} */
    AdComponent.prototype.lightboxData;
    /** @type {?} */
    AdComponent.prototype.events;
}
export class CrystalLightbox {
    /**
     * @param {?} componentFactoryResolver
     * @param {?} appRef
     * @param {?} injector
     */
    constructor(componentFactoryResolver, appRef, injector) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.injector = injector;
    }
    /**
     * @param {?} component
     * @param {?} lightboxData
     * @return {?}
     */
    appendComponentToBody(component, lightboxData) {
        /** @type {?} */
        const componentRef = this.componentFactoryResolver
            .resolveComponentFactory(component)
            .create(this.injector);
        ((/** @type {?} */ (componentRef.instance))).lightboxData = lightboxData;
        this.appRef.attachView(componentRef.hostView);
        /** @type {?} */
        const domElem = (/** @type {?} */ (((/** @type {?} */ (componentRef.hostView))).rootNodes[0]));
        // Add to body
        document.body.appendChild(domElem);
        ((/** @type {?} */ (componentRef.instance))).events.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (event.type === 'close') {
                this.appRef.detachView(componentRef.hostView);
                componentRef.destroy();
            }
        }));
    }
    /**
     * @param {?} lightboxData
     * @return {?}
     */
    open(lightboxData) {
        lightboxData.properties = this.applyPropertieDefaults(DefaultProperties, lightboxData.properties);
        /** @type {?} */
        let component = this.getLightboxComponent();
        this.appendComponentToBody(component, lightboxData);
    }
    /**
     * @return {?}
     */
    getLightboxComponent() {
        return LightboxComponent;
    }
    /**
     * @param {?} defaultProperties
     * @param {?} properties
     * @return {?}
     */
    applyPropertieDefaults(defaultProperties, properties) {
        if (!properties) {
            properties = {};
        }
        if (!properties.index) {
            properties.index = 0;
        }
        this._defaultProperties = Object.assign({}, defaultProperties);
        return Object.assign(this._defaultProperties, properties);
    }
}
CrystalLightbox.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CrystalLightbox.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ApplicationRef },
    { type: Injector }
];
if (false) {
    /** @type {?} */
    CrystalLightbox.prototype.isMobile;
    /** @type {?} */
    CrystalLightbox.prototype._defaultProperties;
    /**
     * @type {?}
     * @private
     */
    CrystalLightbox.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    CrystalLightbox.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    CrystalLightbox.prototype.injector;
}
