// @crystalui/angular-lightbox/esm5/lib/interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function LightboxData() { }
if (false) {
    /** @type {?|undefined} */
    LightboxData.prototype.images;
    /** @type {?|undefined} */
    LightboxData.prototype.image;
    /** @type {?|undefined} */
    LightboxData.prototype.properties;
    /** @type {?|undefined} */
    LightboxData.prototype.index;
}
/**
 * @record
 */
export function Images() { }
/**
 * @record
 */
export function Image() { }
if (false) {
    /** @type {?} */
    Image.prototype.path;
    /** @type {?|undefined} */
    Image.prototype.width;
    /** @type {?|undefined} */
    Image.prototype.height;
}
/**
 * @record
 */
export function ImageExtended() { }
if (false) {
    /** @type {?|undefined} */
    ImageExtended.prototype.fullImage;
    /** @type {?|undefined} */
    ImageExtended.prototype.thumbnailImage;
    /** @type {?|undefined} */
    ImageExtended.prototype.nativeElement;
}
/**
 * @record
 */
export function Properties() { }
if (false) {
    /** @type {?|undefined} */
    Properties.prototype.loop;
    /** @type {?|undefined} */
    Properties.prototype.index;
    /** @type {?|undefined} */
    Properties.prototype.counter;
    /** @type {?|undefined} */
    Properties.prototype.imageMaxHeight;
    /** @type {?|undefined} */
    Properties.prototype.imageMaxWidth;
    /** @type {?|undefined} */
    Properties.prototype.animationDuration;
    /** @type {?|undefined} */
    Properties.prototype.animationMode;
    /** @type {?|undefined} */
    Properties.prototype.animationTimingFunction;
    /** @type {?|undefined} */
    Properties.prototype.closeButtonText;
    /** @type {?|undefined} */
    Properties.prototype.counterSeparator;
    /** @type {?|undefined} */
    Properties.prototype.disable;
    /** @type {?|undefined} */
    Properties.prototype.simpleMode;
    /** @type {?|undefined} */
    Properties.prototype.backgroundColor;
    /** @type {?|undefined} */
    Properties.prototype.backgroundOpacity;
    /** @type {?|undefined} */
    Properties.prototype.hideThumbnail;
    /** @type {?|undefined} */
    Properties.prototype.imagePathError;
    /** @type {?|undefined} */
    Properties.prototype.gestureEnable;
}
/** @type {?} */
export var DefaultProperties = {
    loop: false,
    index: 0,
    counter: false,
    imageMaxHeight: "100%",
    imageMaxWidth: "100%",
    animationDuration: 350,
    animationMode: 'zoom-preloader',
    animationTimingFunction: 'cubic-bezier(0.475, 0.105, 0.445, 0.945)',
    closeButtonText: 'Close',
    counterSeparator: '/',
    disable: false,
    simpleMode: false,
    backgroundColor: 'black',
    backgroundOpacity: 1,
    hideThumbnail: true,
};
