// @crystalui/angular-lightbox/esm5/lib/lightbox.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LightboxCommonComponent } from './lightbox-common.component';
import { LightboxComponent } from './lightbox.component';
import { CrystalLightbox } from './lightbox.service';
import { EventService } from './event.service';
import { LightboxDirective } from './lightbox.directive';
import { LightboxGroupDirective } from './lightbox-group.directive';
var CrystalLightboxModule = /** @class */ (function () {
    function CrystalLightboxModule() {
    }
    CrystalLightboxModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        LightboxCommonComponent,
                        LightboxComponent,
                        LightboxDirective,
                        LightboxGroupDirective
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        LightboxDirective,
                        LightboxGroupDirective
                    ],
                    providers: [
                        CrystalLightbox,
                        EventService
                    ],
                    bootstrap: [],
                    entryComponents: [
                        LightboxComponent
                    ]
                },] }
    ];
    return CrystalLightboxModule;
}());
export { CrystalLightboxModule };
