// @crystalui/angular-lightbox/esm5/lib/lightbox.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver, ApplicationRef } from '@angular/core';
import { LightboxComponent } from './lightbox.component';
import { DefaultProperties } from './default-properties';
/**
 * @record
 */
export function AdComponent() { }
if (false) {
    /** @type {?} */
    AdComponent.prototype.lightboxData;
    /** @type {?} */
    AdComponent.prototype.events;
}
var CrystalLightbox = /** @class */ (function () {
    function CrystalLightbox(componentFactoryResolver, appRef, injector) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.injector = injector;
    }
    /**
     * @param {?} component
     * @param {?} lightboxData
     * @return {?}
     */
    CrystalLightbox.prototype.appendComponentToBody = /**
     * @param {?} component
     * @param {?} lightboxData
     * @return {?}
     */
    function (component, lightboxData) {
        var _this = this;
        /** @type {?} */
        var componentRef = this.componentFactoryResolver
            .resolveComponentFactory(component)
            .create(this.injector);
        ((/** @type {?} */ (componentRef.instance))).lightboxData = lightboxData;
        this.appRef.attachView(componentRef.hostView);
        /** @type {?} */
        var domElem = (/** @type {?} */ (((/** @type {?} */ (componentRef.hostView))).rootNodes[0]));
        // Add to body
        document.body.appendChild(domElem);
        ((/** @type {?} */ (componentRef.instance))).events.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (event.type === 'close') {
                _this.appRef.detachView(componentRef.hostView);
                componentRef.destroy();
            }
        }));
    };
    /**
     * @param {?} lightboxData
     * @return {?}
     */
    CrystalLightbox.prototype.open = /**
     * @param {?} lightboxData
     * @return {?}
     */
    function (lightboxData) {
        lightboxData.properties = this.applyPropertieDefaults(DefaultProperties, lightboxData.properties);
        /** @type {?} */
        var component = this.getLightboxComponent();
        this.appendComponentToBody(component, lightboxData);
    };
    /**
     * @return {?}
     */
    CrystalLightbox.prototype.getLightboxComponent = /**
     * @return {?}
     */
    function () {
        return LightboxComponent;
    };
    /**
     * @param {?} defaultProperties
     * @param {?} properties
     * @return {?}
     */
    CrystalLightbox.prototype.applyPropertieDefaults = /**
     * @param {?} defaultProperties
     * @param {?} properties
     * @return {?}
     */
    function (defaultProperties, properties) {
        if (!properties) {
            properties = {};
        }
        if (!properties.index) {
            properties.index = 0;
        }
        this._defaultProperties = Object.assign({}, defaultProperties);
        return Object.assign(this._defaultProperties, properties);
    };
    CrystalLightbox.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CrystalLightbox.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ApplicationRef },
        { type: Injector }
    ]; };
    return CrystalLightbox;
}());
export { CrystalLightbox };
if (false) {
    /** @type {?} */
    CrystalLightbox.prototype.isMobile;
    /** @type {?} */
    CrystalLightbox.prototype._defaultProperties;
    /**
     * @type {?}
     * @private
     */
    CrystalLightbox.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    CrystalLightbox.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    CrystalLightbox.prototype.injector;
}
